html/user-activation/consumption-sameorigin.html

```diff
@@ -84,25 +84,31 @@
         // Phase switching.
         if (msg.type.endsWith("-loaded")) {
             if (--num_children_to_load == 0)
-                finishLoadPhase();
+                await finishLoadPhase();
         } else if (msg.type.endsWith("-report")) {
             if (--num_children_to_report == 0)
                 finishReportPhase();
         }
     });
+    async function createIframes() {
+        const child1 = document.createElement("iframe");
+        child1.src = "resources/child-one.html";
+        child1.id = "child1";
+        document.body.appendChild(child1);
+        await new Promise((resolve) => (child1.onload = resolve));
```

Comment:
That would be more correct, yes. However, my testing this on a few hundred runs seems to confirm it's fine (all that is needed here is that the child1 loads first).

Reply:
You're misreading that note @mustaqahmed. That's just about event propagation. `iframe`'s load event will fire after its inner document `load` event.

Reply:
Yeah, that's correct.

Reply:
So yes, but that's "too magical". I prefer the explicit loading order of the iframes. 

The problem right now is that the messages are coming in out of sync because the iframes are sending messages depending on when they load.  

I'm open to dropping `createIframes()`, but I'm not understanding how using window.onload simplifies things when it's the underlying iframes that are causing the issue. 

The only way around it would be to postMessage to each frame from the top frame that it's ok to start sending messages, right? Or am I missing something obvious?

Reply:
Yeah, it's no different from the top-level document `load` event. It generally waits for all subresources to load.

Reply:
Ok, yeah, I didn't want to touch the logic of the underlying iframes. But yes, as they are same-origin, all the messages can go away.

Reply:
@mustaqahmed, can you help me understand what cause the consumption of the user activation here?  

I'd still prefer not to rewrite the test if I can avoid it. 

However, if I flatten the test to:

```JS
      async function runTest() {
        promise_test(async (t) => {
            const child1 = document.getElementById("child1");
            const child1Activation = child1.contentWindow.navigator.userActivation;
            const child2 = document.getElementById("child2");
            const child2Activation = child2.contentWindow.navigator.userActivation;
            const grandChild = child2.contentDocument.getElementById("grandchild");
            const grandChildActivation = grandChild.contentWindow.navigator.userActivation;
            
            assert_false(navigator.userActivation.isActive, "Parent frame isActive initial state");
            assert_false(navigator.userActivation.hasBeenActive, "Parent frame hasBeenActive initial state");
            assert_false(child1Activation.isActive, "Child1 frame isActive initial state");
            assert_false(child1Activation.hasBeenActive, "Child1 frame hasBeenActive initial state");
            assert_false(child2Activation.isActive, "Child2 frame isActive initial state");
            assert_false(child2Activation.hasBeenActive, "Child2 frame hasBeenActive initial state");
            assert_false(grandChildActivation.isActive, "Grandchild frame isActive initial state");
            assert_false(grandChildActivation.hasBeenActive, "Grandchild frame hasBeenActive initial state");
            
            await test_driver.click("click", child2);
             
            // and so on... 
        });
      }
``` 

I get different results? I can't see anything in the iframes that consume the user activation?

Reply:
> I get different results? I can't see anything in the iframes that consume the user activation?

We shouldn't get a different result, because a [consumption call](https://html.spec.whatwg.org/multipage/interaction.html#consume-user-activation) affects the whole frame tree.

Could you please elaborate which assertions are different between flattened vs un-flattened frame load code?

Is there a chance test_driver implementation is causing the flake here?  This same-origin test is not flaky in Chrome's internal bots, but the cross-origin one is flaky and we can't fix that because of a test_driver impl limitation.

In any case, I think a virtual meeting would be more efficient here to share our thoughts.  Let's plan one.

Reply:
> We shouldn't get a different result, because a [consumption call](https://html.spec.whatwg.org/multipage/interaction.html#consume-user-activation) affects the whole frame tree.

Oh, wait! But the test again does something non-standard:

```JS
window.open().close();
```

That shouldn't consume the user activation: at least, there is nothing in HTML saying that `window.close()` consumes it. 

The flattened version should be this (all false, click(), all true):

```JS
async function runTest() {
  promise_test(async (t) => {
      const child1 = document.getElementById("child1");
      const child1Activation = child1.contentWindow.navigator.userActivation;
      const child2 = document.getElementById("child2");
      const child2Activation = child2.contentWindow.navigator.userActivation;
      const grandChild = child2.contentDocument.getElementById("grandchild");
      const grandChildActivation = grandChild.contentWindow.navigator.userActivation;
      
      assert_false(navigator.userActivation.isActive, "Parent frame isActive initial state");
      assert_false(navigator.userActivation.hasBeenActive, "Parent frame hasBeenActive initial state");
      assert_false(child1Activation.isActive, "Child1 frame isActive initial state");
      assert_false(child1Activation.hasBeenActive, "Child1 frame hasBeenActive initial state");
      assert_false(child2Activation.isActive, "Child2 frame isActive initial state");
      assert_false(child2Activation.hasBeenActive, "Child2 frame hasBeenActive initial state");
      assert_false(grandChildActivation.isActive, "Grandchild frame isActive initial state");
      assert_false(grandChildActivation.hasBeenActive, "Grandchild frame hasBeenActive initial state");
      
      await test_driver.bless("click", null, child1.contentWindow);

      assert_true(navigator.userActivation.isActive, "Parent frame isActive after click");
      assert_true(navigator.userActivation.hasBeenActive, "Parent frame hasBeenActive after click");
      assert_true(child1Activation.isActive, "Child1 frame isActive after click");
      assert_true(child1Activation.hasBeenActive, "Child1 frame hasBeenActive after click");
      assert_true(child2Activation.isActive, "Child2 frame isActive after click");
      assert_true(child2Activation.hasBeenActive, "Child2 frame hasBeenActive after click");
      assert_true(grandChildActivation.isActive, "Grandchild frame isActive after click");
      assert_true(grandChildActivation.hasBeenActive, "Grandchild frame hasBeenActive after click");
  });
}
```

Reply:
> In any case, I think a virtual meeting would be more efficient here to share our thoughts. Let's plan one.

Happy to chat, but I think the biggest points of contention are that we don't have a consistent way of consuming the user activation. 

So far, the tests have used two non-standard ways to try to consume it: 

 1. requestFullscreen() + exitFullscreen()
 2. window.open() + window.close() 

We should also avoid `window.open()` tests. They don't work well on mobile devices. 

I'm thinking we should remove or "-tentative" such tests until we have an actual solution to prevent codifying non-standard behavior.

Reply:
https://dontcallmedom.github.io/webdex/c.html#consume%20user%20activation%40%40html%25%25dfn might help